File: Methods/Java/Methods/Matrix/TotalPivot.java
package Methods.Matrix;

import Errors.MatrixWithColumnZero;
import Errors.NoDimensionMatrix;
import Errors.NotSquareMatrix;
import Share.MatrixUtil;

public class TotalPivot extends PartialPivot{

    private int[] positionStamp;

    @Override
    public double[] execute(double[][] matrix, double[] b) throws NotSquareMatrix, NoDimensionMatrix, MatrixWithColumnZero {

        initPositionStamp(matrix.length);
        return sort(super.execute(matrix, b), positionStamp);
    }

    void pivot(double[][] matrix, int index, double[] b) {
        int row = index;
        int col = index;

        for (int i = row; i < matrix.length; ++i) {
            for (int j = col; j < matrix.length; ++j) {
                if (Math.abs(matrix[i][j]) > Math.abs(matrix[row][col])) {
                    row = i;
                    col = j;
                }
            }
        }
        if (col != index) {
            MatrixUtil.swapColumns(matrix, col, index);
            MatrixUtil.swapValues(positionStamp, col, index);
        }
        if (row != index) {
            MatrixUtil.swapRows(matrix, row, index);
            MatrixUtil.swapValues(b, row, index);
        }
    }

    private void initPositionStamp(int length) {
        if (positionStamp == null) {
            positionStamp = new int[length];
            for (int i = 0; i < length; i++) {
                positionStamp[i] = i;
            }
        }
    }

    private double[] sort(double[] values, int[] positions) {
        double[] sortedValues = new double[values.length];

        for (int i = 0; i < positions.length; ++i)
            sortedValues[positions[i]] = values[i];

        return sortedValues;
    }
}

File: Methods/Python/TotalPivot.py
import MatrixUtils as mu
import Gauss as gs

def gauss(matrix, b):
    positionStamp = [0] * len(matrix)

    for i in range(len(positionStamp)):
        positionStamp[i] = i

    for i in range(len(matrix)):
        pivot(matrix, i, b, positionStamp)

        for j in range(i+1,len(matrix)):
            multiplicand = matrix[j][i] / matrix[i][i]
            gs.elimination(i, j, multiplicand, matrix, b)
    return sort(gs.backwardSubstitution(matrix, b), positionStamp)


def pivot(matrix, index, b, positionStamp):
    row = index
    col = index

    for i in range(row, len(matrix)):
        for j in range(col, len(matrix)):
            if abs(matrix[i][j]) > abs(matrix[row][col]):
                row = i
                col = j
    if col != index:
        mu.swapCols(matrix, col, index)
        mu.swapValues(positionStamp, col, index)
    if row != index:
        mu.swapRows(matrix, row, index)
        mu.swapValues(b, row, index)

def sort(values, positions):
    sortedValues = [0] * len(values)
    for i in range(len(positions)):
        sortedValues[positions[i]] = values[i]
    return sortedValues



a = [   [ 2, -1, 2],
        [ 1,  1, 1],
        [-1,  4, 1]
    ]

b = [4,2,3]
print(gauss(a,b))